src/native/libs/System.Security.Cryptography.Native/pal_evp_kdf.c

```diff
@@ -157,3 +157,166 @@ int32_t CryptoNative_KbkdfHmacOneShot(
+    assert(key != NULL || keyLength == 0);
+    assert(keyLength >= 0);
+    assert(algorithm);
+    assert(destination);
+    assert(destinationLength > 0);
+    assert(salt != NULL || saltLength == 0);
+    assert(info != NULL || infoLength == 0);
+
+    ERR_clear_error();
+
+#ifdef NEED_OPENSSL_3_0
+    if (API_EXISTS(EVP_KDF_CTX_new))
+    {
+        assert(API_EXISTS(EVP_KDF_CTX_free));
+        assert(API_EXISTS(EVP_KDF_derive));
+        assert(API_EXISTS(OSSL_PARAM_construct_utf8_string));
+        assert(API_EXISTS(OSSL_PARAM_construct_octet_string));
+        assert(API_EXISTS(OSSL_PARAM_construct_end));
+
+        EVP_KDF_CTX* ctx = EVP_KDF_CTX_new(kdf);
+        int32_t ret = 0;
+
+        if (ctx == NULL)
+        {
+            goto cleanup;
+        }
+
+        size_t keyLengthT = Int32ToSizeT(keyLength);
+        size_t destinationLengthT = Int32ToSizeT(destinationLength);
+        size_t saltLengthT = Int32ToSizeT(saltLength);
+        size_t infoLengthT = Int32ToSizeT(infoLength);
+
+        OSSL_PARAM params[] =
+        {
+            OSSL_PARAM_construct_octet_string(OSSL_KDF_PARAM_KEY, (void*)key, keyLengthT),
+            OSSL_PARAM_construct_utf8_string(OSSL_KDF_PARAM_DIGEST, algorithm, 0),
+            OSSL_PARAM_construct_octet_string(OSSL_KDF_PARAM_SALT, (void*)salt, saltLengthT),
+            OSSL_PARAM_construct_octet_string(OSSL_KDF_PARAM_INFO, (void*)info, infoLengthT),
+            OSSL_PARAM_construct_utf8_string(OSSL_KDF_PARAM_MODE, operation, 0),
+            OSSL_PARAM_construct_end(),
+        };
+
+        if (EVP_KDF_derive(ctx, destination, destinationLengthT, params) <= 0)
+        {
+            goto cleanup;
+        }
+
+        ret = 1;
+
+cleanup:
+        if (ctx != NULL)
+        {
+            EVP_KDF_CTX_free(ctx);
+        }
+
+        return ret;
+    }
+#else
+    (void)kdf;
+    (void)operation;
+    (void)key;
+    (void)keyLength;
+    (void)algorithm;
+    (void)salt;
+    (void)saltLength;
+    (void)info;
+    (void)infoLength;
+    (void)destination;
+    (void)destinationLength;
+    assert(0 && "Inconsistent EVP_KDF API availability.");
+#endif
+    return 0;
+}
+
+int32_t CryptoNative_HkdfDeriveKey(
+    EVP_KDF* kdf,
+    uint8_t* ikm,
+    int32_t ikmLength,
+    char* algorithm,
+    uint8_t* salt,
+    int32_t saltLength,
+    uint8_t* info,
+    int32_t infoLength,
+    uint8_t* destination,
+    int32_t destinationLength)
+{
+    return HkdfCore(
+        kdf,
+        "EXTRACT_AND_EXPAND",
+        ikm,
+        ikmLength,
+        algorithm,
+        salt,
+        saltLength,
+        info,
+        infoLength,
+        destination,
+        destinationLength);
+}
+
+int32_t CryptoNative_HkdfExpand(
+    EVP_KDF* kdf,
+    uint8_t* prk,
+    int32_t prkLength,
+    char* algorithm,
+    uint8_t* info,
+    int32_t infoLength,
+    uint8_t* destination,
+    int32_t destinationLength)
+{
+    return HkdfCore(
+        kdf,
+        "EXPAND_ONLY",
+        prk,
```

Comment:
I like the containment of the string literals here.

I do wonder if we should be static_asserting that they're equal to a named constant, or doing the ifdef / assert / else / define pattern so we can use it here symbolically

Reply:
The strings as far as I can tell have no named constants. They are documented magic strings.

It does have constants that are integers. So I switched to that and did all the define assert stuff.